### PR TITLE
Comparing matrix with object

### LINKED ISSUE #19361: Comparing Matrix with object() raises
This should just give False:
```julia
In [5]: Matrix([[1]]) == object()                                                                                                 
/Users/enojb/current/sympy/sympy/sympy/core/sympify.py:383: SymPyDeprecationWarning: 

String fallback in sympify has been deprecated since SymPy 1.6. Use
sympify(str(obj)) or sympy.core.sympify.converter or obj._sympy_
instead. See https://github.com/sympy/sympy/issues/18066 for more
info.

  SymPyDeprecationWarning(
---------------------------------------------------------------------------
SyntaxError                               Traceback (most recent call last)
~/current/sympy/sympy/sympy/core/sympify.py in sympify(a, locals, convert_xor, strict, rational, evaluate)
    405         a = a.replace('\n', '')
--> 406         expr = parse_expr(a, local_dict=locals, transformations=transformations, evaluate=evaluate)
    407     except (TokenError, SyntaxError) as exc:

~/current/sympy/sympy/sympy/parsing/sympy_parser.py in parse_expr(s, local_dict, transformations, global_dict, evaluate)
   1007 
-> 1008     return eval_expr(code, local_dict, global_dict)
   1009 

~/current/sympy/sympy/sympy/parsing/sympy_parser.py in eval_expr(code, local_dict, global_dict)
    901     """
--> 902     expr = eval(
    903         code, global_dict, local_dict)  # take local objects in preference

SyntaxError: invalid syntax (<string>, line 1)

During handling of the above exception, another exception occurred:

SympifyError                              Traceback (most recent call last)
<ipython-input-5-0e1448e70140> in <module>
----> 1 Matrix([[1]]) == object()

~/current/sympy/sympy/sympy/matrices/dense.py in __eq__(self, other)
     44 
     45     def __eq__(self, other):
---> 46         other = sympify(other)
     47         self_shape = getattr(self, 'shape', None)
     48         other_shape = getattr(other, 'shape', None)

~/current/sympy/sympy/sympy/core/sympify.py in sympify(a, locals, convert_xor, strict, rational, evaluate)
    406         expr = parse_expr(a, local_dict=locals, transformations=transformations, evaluate=evaluate)
    407     except (TokenError, SyntaxError) as exc:
--> 408         raise SympifyError('could not parse %r' % a, exc)
    409 
    410     return expr

SympifyError: Sympify of expression 'could not parse '<object object at 0x1130206c0>'' failed, because of exception being raised:
SyntaxError: invalid syntax (<string>, line 1)
```
If sympify is used in `__eq__` then the exception needs to be caught and `NotImplemented` should be returned. The `sympify_return` decorator used for special methods in `Expr` is also applicable here.

Similar problems apply to other `__eq__` methods. They will raise when sympify/conversion fails but that shouldn't happen (especially with `__eq__`).

### PR DESCRIPTION
**Pull request by sidhu1012**
#### References to other Issues or PRs

Fixes #19361

#### Brief description of what is fixed or changed
Corrected __eq__() of Matrix

#### Release Notes

- matrices
       -  Changed behaviour of __eq__() of class DenseMatrix

### CODE REVIEW

**Comment by sidhu1012** [comment]
@oscarbenjamin Does it seems right?

**Comment by friyaz** [comment]
Always add test cases that will fail without the changes you make. This ensures that the problem does not occur in the future.

**Comment by oscarbenjamin** [comment]
There is a decorator `sympify_return` for sympifying arguments or returning NotImplemented. It is used in `sympy/core/expr.py`.

**Comment by sidhu1012** [comment]
> Always add test cases that will fail without the changes you make. This ensures that the problem does not occur in the future.

Done

**Comment by sidhu1012** [comment]
> There is a decorator `sympify_return` for sympifying arguments or returning NotImplemented. It is used in `sympy/core/expr.py`.

Used it , decorator really easied it.

**Comment by sidhu1012** [comment]
@oscarbenjamin pls review

File: sympy/matrices/tests/test_densearith.py [hunk 1]
```diff
@@ -52,3 +54,7 @@ def test_mulmatscaler():
 
     assert mulmatscaler(a, ZZ(4), ZZ) == [[ZZ(4), ZZ(0), ZZ(0)], [ZZ(0), ZZ(4), ZZ(0)], [ZZ(0), ZZ(0), ZZ(4)]]
     assert mulmatscaler(b, ZZ(1), ZZ) == [[ZZ(3), ZZ(7), ZZ(4)], [ZZ(2), ZZ(4), ZZ(5)], [ZZ(6), ZZ(2), ZZ(3)]]
+
+def test_eq():
+    assert Matrix([[1]]) != object() #Earlier gave Syntax and SympifyError , now returns False
```
Comment (friyaz) [inline sympy/matrices/tests/test_densearith.py:59]: These comments seem unnecessary to me.

Comment (sidhu1012) [reply sympy/matrices/tests/test_densearith.py:59]: I'll remove them

Comment (friyaz) [reply sympy/matrices/tests/test_densearith.py:59]: If you want to comment, you can reword it to make them more clear. But I think it is better to remove them.

Comment (sidhu1012) [reply sympy/matrices/tests/test_densearith.py:59]: Sure

**Comment by sidhu1012** [comment]
@friyaz Removed comments. Does everything else seems fine?

**Comment by friyaz** [comment]
Looks good to me.

**Comment by sidhu1012** [comment]
> > There is a decorator `sympify_return` for sympifying arguments or returning NotImplemented. It is used in `sympy/core/expr.py`.

@oscarbenjamin  @friyaz  dont think using decorator is a good way because it is making other test fail

**Comment by sidhu1012** [comment]
> @oscarbenjamin  @friyaz  dont think using decorator is a good way because it is making other test fail.
Is something missing in this PR or PR's first commit is a better solve for this.

**Comment by sidhu1012** [comment]
@oscarbenjamin So shall I drop the decorator and use the method of 1st commit?

**Comment by sidhu1012** [comment]
This way is better as it doesn't break anything. Decorator breaks the code

**Comment by sidhu1012** [comment]
@oscarbenjamin Does it seems right now??

**Comment by oscarbenjamin** [comment]
Why didn't the decorator work?

**Comment by sidhu1012** [comment]
> Why didn't the decorator work?

Decorator broke the code , and gave way too many failures

**Comment by oscarbenjamin** [comment]
Can you give an example of something that didn't work?

**Comment by sidhu1012** [comment]
> Can you give an example of something that didn't work?

https://travis-ci.org/github/sympy/sympy/builds/732235398

See the build at travis , all 4 test failed, test files and doc tests.

**Comment by sidhu1012** [comment]
@oscarbenjamin Is it good without decorator?

**Comment by oscarbenjamin** [comment]
Yes, it should be fine but we should check why it didn't work with the decorator. Maybe the decorator can be improved or maybe it's just not suitable for this. Without investigating the failures we won't know.

**Comment by sidhu1012** [comment]
> Yes, it should be fine but we should check why it didn't work with the decorator. Maybe the decorator can be improved or maybe it's just not suitable for this. Without investigating the failures we won't know.

Okay , i'll look into it as much as I can

**Comment by sidhu1012** [comment]
> or maybe it's just not suitable for this. 

The number of test failures suggest decorator isn't suitable for this. Because the number of failures are really high and are in almost every module

**Comment by sidhu1012** [comment]
@oscarbenjamin What do you say? Is is good to go?

**Comment by oscarbenjamin** [comment]
Why doesn't the decorator work?

The error from the test logs is
```
    if not isinstance(other, expectedcls):
TypeError: isinstance() arg 2 must be a type or tuple of types
```
That suggests that the decorator wasn't used correctly.

**Comment by sidhu1012** [comment]
I'll give it another try

**Comment by sidhu1012** [comment]
@oscarbenjamin Please review now

**Comment by sidhu1012** [comment]
@oscarbenjamin Review please. I think it's good to go now.

**Comment by sidhu1012** [comment]
@oscarbenjamin Is it good to go now??

File: sympy/matrices/dense.py [hunk 2]
```diff
@@ -35,15 +36,16 @@ def _compare_sequence(a, b):
...
 
     _op_priority = 10.01
     _class_priority = 4
 
+    @sympify_return([('other', 'DenseMatrix')], NotImplemented)
     def __eq__(self, other):
-        other = sympify(other)
         self_shape = getattr(self, 'shape', None)
         other_shape = getattr(other, 'shape', None)
         if None in (self_shape, other_shape):
```
Comment (oscarbenjamin) [inline sympy/matrices/dense.py:51]: The conditions below suggest that sometimes `other` is not a `DenseMatrix`. Is that not ever the case?

Comment (sidhu1012) [reply sympy/matrices/dense.py:51]: It works for all cases .

**Comment by sidhu1012** [comment]
@oscarbenjamin I have checked and updates test file with other Matrix types also. It works as expected.

**Comment by sidhu1012** [comment]
@oscarbenjamin I have checked and updates test file with other Matrix types also. It works as expected.

**Comment by sidhu1012** [comment]
@oscarbenjamin Please review now

**Comment by sidhu1012** [comment]
Since @oscarbenjamin is busy with 1.7 update. Can someone else review please.

**Comment by sidhu1012** [comment]
Can someone please review

**Review by friyaz — APPROVED** [review]

**Comment by sidhu1012** [comment]
@sylee957 Please review.

**Comment by sidhu1012** [comment]
Since no one is reviewing, @oscarbenjamin please review it when free

**Comment by sidhu1012** [comment]
Can someone please review this PR.

File: sympy/matrices/dense.py [hunk 3]
```diff
@@ -35,15 +36,16 @@ def _compare_sequence(a, b):
...
 
+@sympify_method_args
 class DenseMatrix(MatrixBase):
 
     is_MatrixExpr = False  # type: bool
 
     _op_priority = 10.01
     _class_priority = 4
 
+    @sympify_return([('other', 'DenseMatrix')], NotImplemented)
```
Comment (sylee957) [inline sympy/matrices/dense.py:47]: Can't we add this to SparseMatrix such that it can be symmetric change?

Comment (sidhu1012) [reply sympy/matrices/dense.py:47]: okay , would do that.

Comment (sidhu1012) [reply sympy/matrices/dense.py:47]: The decorator can be  used when sympifying code, since the SparseMatrix isn't sympifying , it shouldn't be added in SparseMatrix

Comment (sylee957) [reply sympy/matrices/dense.py:47]: I'd rather change that SparseMatrix to sympify the arguments before calls
It is problematic if operations supported in DenseMatrix is not supported by SparseMatrix only because of sympify
For example,
```
from sympy import *
import mpmath

m = mpmath.matrix([[1, 2], [3, 4]])
Matrix([[1, 2], [3, 4]]) == m # True
SparseMatrix([[1, 2], [3, 4]]) == m # False

**Comment by sidhu1012** [comment]
@sylee957 @oscarbenjamin I have removed the decorator and user _sympify instead .
@oscarbenjamin was right https://github.com/sympy/sympy/pull/20175#discussion_r502523897 , the dexorator would make Dense Matrix == Dense only and same goes to sparse, earlier it was working because changes weren't symmetric.

**Comment by oscarbenjamin** [comment]
@sidhu1012 we are having problems with CI (#20289). The tests are unlikely to pass for these PRs until that is resolved. I cancelled the previous run of the tests because because they were slowing down the tests for #20297

**Comment by sidhu1012** [comment]
@oscarbenjamin Okay , please review this PR though

**Comment by sidhu1012** [comment]
@oscarbenjamin @sylee957  Please review
